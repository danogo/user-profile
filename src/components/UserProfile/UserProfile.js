import React from 'react';
import styles from './UserProfile.module.css';
import Header from './Header/Header';
import Comments from './Comments/Comments';
import Modal from './Modal/Modal';
const comments = require('../../data.json');

class UserProfile extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      likes: 121,
      following: 723,
      followers: 4433,
      comments: [],
      isLiked: false,
      isFollowed: false,
      showComments: true,
      showModal: false,
      addedComment: false,
    };
  }

  componentDidMount() {
    this.setState({
      comments: comments
    });
  }

  componentDidUpdate(prevProps, prevState) {
    if((prevState.addedComment !== this.state.addedComment) && this.state.showComments) {
      document.getElementById('inputComment').value = '';
      setTimeout(() => {
        const list = document.getElementById('comList');
        list.lastElementChild.scrollIntoView({behavior: 'smooth'});
      }, prevState.showComments ? 0 : 400);
    }
    if(this.state.showModal) {
      const inputModal = document.getElementById('inputModal');
      inputModal.focus();
    }
  }

  handleHeartClick = () => {
    if(!this.state.isLiked) {
      this.setState({
        likes: this.state.likes + 1,
        isLiked: !this.state.isLiked
      })
    } else {
      this.setState({
        likes: this.state.likes -1,
        isLiked: !this.state.isLiked
      })
    }
  }

  handleFollowClick = () => {
    if(!this.state.isFollowed) {
      this.setState({
        followers: this.state.followers + 1,
        isFollowed: !this.state.isFollowed
      })
    } else {
      this.setState({
        followers: this.state.followers -1,
        isFollowed: !this.state.isFollowed
      })
    }
  }

  addComment = event => {
    // grabbing pressed key
    let pressed = event.key || event.keyCode;
    // create new comment if user pressed enter and input is not empty
    if ((pressed === 'Enter' || pressed === 13) && event.target.value) {
      const newComment = {
        username: 'John Doe',
        date:  Date.now(),
        content: event.target.value
      } 
      // update comments array based on previous state
      this.setState(prevState => ({
        comments: [...prevState.comments, newComment],
        showComments: true,
        addedComment: !prevState.addedComment
      }));
    }
  }

  toggleComments = () => {
    this.setState(prevState => ({
      showComments: !prevState.showComments
    }));
  }

  handleModalShow = () => {
    this.setState({
      showModal: true
    });
  }

  handleModalHide = () => {
    this.setState({
      showModal: false
    });
  }
  
  render() {
    return (
      <div className={styles.container}>
        <Header onHeartClick={this.handleHeartClick} onFollowClick={this.handleFollowClick} onShareClick={this.handleModalShow} likes={this.state.likes} following={this.state.following} followers={this.state.followers} isFollowed={this.state.isFollowed} isLiked={this.state.isLiked}/>
        <Comments onInput={this.addComment} onToggleComments={this.toggleComments} comments={this.state.comments}  showComments={this.state.showComments}/>
        <Modal onBtnClick={this.handleModalHide} showModal={this.state.showModal}/>
      </div>
    );
  }
}

export default UserProfile;