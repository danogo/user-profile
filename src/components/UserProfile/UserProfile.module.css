@font-face {
  font-family: 'Montserrat';
  font-style: normal;
  font-weight: 400;
  src: local('Montserrat Regular'), local('Montserrat-Regular'), url(https://fonts.gstatic.com/s/montserrat/v12/JTUSjIg1_i6t8kCHKm459Wlhyw.woff2) format('woff2');
  unicode-range: U+0000-00FF, U+0131, U+0152-0153, U+02BB-02BC, U+02C6, U+02DA, U+02DC, U+2000-206F, U+2074, U+20AC, U+2122, U+2191, U+2193, U+2212, U+2215, U+FEFF, U+FFFD;
}

.container {
  width: 90%;
  max-width: 500px;
  margin: 0 auto;
  background-color: #FFFFFF;
  box-shadow: 0 0 4px #ACACAC50;
  position: relative;
  border-radius: 5px;
  font-family: 'Montserrat', Helvetica, Arial, sans-serif;
  font-weight: 400;
  line-height: 1.7;
  padding-bottom: 3rem;
}

.container::before {
  content: '';
  display: block;
  position: absolute;
  top: 0;
  width: 100%;
  background-color: #002C71;
  height: 12%;
  min-height: 7.775rem;
  border-radius: 5px 5px 0 0;
}

@media only screen and (max-width: 550px) {
  .container::before {
    min-height: 8.411rem;
  }
}