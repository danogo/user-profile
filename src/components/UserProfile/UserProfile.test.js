import React from 'react';
import { shallow } from 'enzyme';
import UserProfile from './UserProfile';

describe('UserProfile', () => {

  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<UserProfile/>);
  });

  test('component renders correctly', () => {
    expect(shallow(<UserProfile/>)).toMatchSnapshot();
  });

  test('renders div container', () => {
    const divs = wrapper.find('div');
    expect(divs.length).toBeGreaterThan(0);
  })

  test('contains all children components', () => {
    const div = wrapper.find('div').first();
    // Enzyme omits the outermost node when using the .children() so we can compare its children against first div's children, which should be the same
    expect(div.children()).toEqual(wrapper.children());
    expect(wrapper.children().length).toBe(3);
  })

  test('doesn\'t receive any props', () => {
    expect(wrapper.instance().props).toEqual({});
  })

  test('profile isn\'t followed or liked by default', () => {
    expect(wrapper.state().isFollowed).toBe(false);
    expect(wrapper.state().isLiked).toBe(false);
  })
});