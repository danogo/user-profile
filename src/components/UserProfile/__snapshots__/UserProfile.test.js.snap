// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserProfile component renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <UserProfile />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Header
          followers={4433}
          following={723}
          isFollowed={false}
          isLiked={false}
          likes={121}
          onFollowClick={[Function]}
          onHeartClick={[Function]}
          onShareClick={[Function]}
        />,
        <Comments
          comments={
            Array [
              Object {
                "content": "Lorem ipsum dolor sit amet consectetur adipisicing elit. Architecto autem nam quae delectus maxime inventore odio libero provident obcaecati a.",
                "date": "1529088547946",
                "username": "Mike Ross",
              },
              Object {
                "content": "Lorem ipsum dolor sit amet consectetur adipisicing elit. Architecto autem nam quae delectus maxime inventore odio libero provident obcaecati a.",
                "date": "1536088547946",
                "username": "Rachel Zein",
              },
              Object {
                "content": "Lorem ipsum dolor sit amet consectetur adipisicing elit. Architecto autem nam quae delectus maxime inventore odio libero provident obcaecati a.",
                "date": "1537998547946",
                "username": "Louis Litt",
              },
              Object {
                "content": "Lorem ipsum dolor sit amet consectetur adipisicing elit. Architecto autem nam quae delectus maxime inventore odio libero provident obcaecati a.",
                "date": "1539358547946",
                "username": "Richard Dawkins",
              },
              Object {
                "content": "Lorem ipsum dolor sit amet consectetur adipisicing elit. Architecto autem nam quae delectus maxime inventore odio libero provident obcaecati a.",
                "date": "1538288547946",
                "username": "Carl Sagan",
              },
              Object {
                "content": "Lorem ipsum dolor sit amet consectetur adipisicing elit. Architecto autem nam quae delectus maxime inventore odio libero provident obcaecati a.",
                "date": "1538998547946",
                "username": "Neil deGrasse Tyson",
              },
              Object {
                "content": "Lorem ipsum dolor sit amet consectetur adipisicing elit. Architecto autem nam quae delectus maxime inventore odio libero provident obcaecati a.",
                "date": "1539488547946",
                "username": "Christopher Hitchens",
              },
              Object {
                "content": "Lorem ipsum dolor sit amet consectetur adipisicing elit. Architecto autem nam quae delectus maxime inventore odio libero provident obcaecati a.",
                "date": "1538700547946",
                "username": "Richard Feynman",
              },
            ]
          }
          onInput={[Function]}
          onToggleComments={[Function]}
          showComments={true}
        />,
        <Modal
          onBtnClick={[Function]}
          showModal={false}
        />,
      ],
      "className": "container",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "followers": 4433,
          "following": 723,
          "isFollowed": false,
          "isLiked": false,
          "likes": 121,
          "onFollowClick": [Function],
          "onHeartClick": [Function],
          "onShareClick": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "comments": Array [
            Object {
              "content": "Lorem ipsum dolor sit amet consectetur adipisicing elit. Architecto autem nam quae delectus maxime inventore odio libero provident obcaecati a.",
              "date": "1529088547946",
              "username": "Mike Ross",
            },
            Object {
              "content": "Lorem ipsum dolor sit amet consectetur adipisicing elit. Architecto autem nam quae delectus maxime inventore odio libero provident obcaecati a.",
              "date": "1536088547946",
              "username": "Rachel Zein",
            },
            Object {
              "content": "Lorem ipsum dolor sit amet consectetur adipisicing elit. Architecto autem nam quae delectus maxime inventore odio libero provident obcaecati a.",
              "date": "1537998547946",
              "username": "Louis Litt",
            },
            Object {
              "content": "Lorem ipsum dolor sit amet consectetur adipisicing elit. Architecto autem nam quae delectus maxime inventore odio libero provident obcaecati a.",
              "date": "1539358547946",
              "username": "Richard Dawkins",
            },
            Object {
              "content": "Lorem ipsum dolor sit amet consectetur adipisicing elit. Architecto autem nam quae delectus maxime inventore odio libero provident obcaecati a.",
              "date": "1538288547946",
              "username": "Carl Sagan",
            },
            Object {
              "content": "Lorem ipsum dolor sit amet consectetur adipisicing elit. Architecto autem nam quae delectus maxime inventore odio libero provident obcaecati a.",
              "date": "1538998547946",
              "username": "Neil deGrasse Tyson",
            },
            Object {
              "content": "Lorem ipsum dolor sit amet consectetur adipisicing elit. Architecto autem nam quae delectus maxime inventore odio libero provident obcaecati a.",
              "date": "1539488547946",
              "username": "Christopher Hitchens",
            },
            Object {
              "content": "Lorem ipsum dolor sit amet consectetur adipisicing elit. Architecto autem nam quae delectus maxime inventore odio libero provident obcaecati a.",
              "date": "1538700547946",
              "username": "Richard Feynman",
            },
          ],
          "onInput": [Function],
          "onToggleComments": [Function],
          "showComments": true,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "onBtnClick": [Function],
          "showModal": false,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Header
            followers={4433}
            following={723}
            isFollowed={false}
            isLiked={false}
            likes={121}
            onFollowClick={[Function]}
            onHeartClick={[Function]}
            onShareClick={[Function]}
          />,
          <Comments
            comments={
              Array [
                Object {
                  "content": "Lorem ipsum dolor sit amet consectetur adipisicing elit. Architecto autem nam quae delectus maxime inventore odio libero provident obcaecati a.",
                  "date": "1529088547946",
                  "username": "Mike Ross",
                },
                Object {
                  "content": "Lorem ipsum dolor sit amet consectetur adipisicing elit. Architecto autem nam quae delectus maxime inventore odio libero provident obcaecati a.",
                  "date": "1536088547946",
                  "username": "Rachel Zein",
                },
                Object {
                  "content": "Lorem ipsum dolor sit amet consectetur adipisicing elit. Architecto autem nam quae delectus maxime inventore odio libero provident obcaecati a.",
                  "date": "1537998547946",
                  "username": "Louis Litt",
                },
                Object {
                  "content": "Lorem ipsum dolor sit amet consectetur adipisicing elit. Architecto autem nam quae delectus maxime inventore odio libero provident obcaecati a.",
                  "date": "1539358547946",
                  "username": "Richard Dawkins",
                },
                Object {
                  "content": "Lorem ipsum dolor sit amet consectetur adipisicing elit. Architecto autem nam quae delectus maxime inventore odio libero provident obcaecati a.",
                  "date": "1538288547946",
                  "username": "Carl Sagan",
                },
                Object {
                  "content": "Lorem ipsum dolor sit amet consectetur adipisicing elit. Architecto autem nam quae delectus maxime inventore odio libero provident obcaecati a.",
                  "date": "1538998547946",
                  "username": "Neil deGrasse Tyson",
                },
                Object {
                  "content": "Lorem ipsum dolor sit amet consectetur adipisicing elit. Architecto autem nam quae delectus maxime inventore odio libero provident obcaecati a.",
                  "date": "1539488547946",
                  "username": "Christopher Hitchens",
                },
                Object {
                  "content": "Lorem ipsum dolor sit amet consectetur adipisicing elit. Architecto autem nam quae delectus maxime inventore odio libero provident obcaecati a.",
                  "date": "1538700547946",
                  "username": "Richard Feynman",
                },
              ]
            }
            onInput={[Function]}
            onToggleComments={[Function]}
            showComments={true}
          />,
          <Modal
            onBtnClick={[Function]}
            showModal={false}
          />,
        ],
        "className": "container",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "followers": 4433,
            "following": 723,
            "isFollowed": false,
            "isLiked": false,
            "likes": 121,
            "onFollowClick": [Function],
            "onHeartClick": [Function],
            "onShareClick": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "comments": Array [
              Object {
                "content": "Lorem ipsum dolor sit amet consectetur adipisicing elit. Architecto autem nam quae delectus maxime inventore odio libero provident obcaecati a.",
                "date": "1529088547946",
                "username": "Mike Ross",
              },
              Object {
                "content": "Lorem ipsum dolor sit amet consectetur adipisicing elit. Architecto autem nam quae delectus maxime inventore odio libero provident obcaecati a.",
                "date": "1536088547946",
                "username": "Rachel Zein",
              },
              Object {
                "content": "Lorem ipsum dolor sit amet consectetur adipisicing elit. Architecto autem nam quae delectus maxime inventore odio libero provident obcaecati a.",
                "date": "1537998547946",
                "username": "Louis Litt",
              },
              Object {
                "content": "Lorem ipsum dolor sit amet consectetur adipisicing elit. Architecto autem nam quae delectus maxime inventore odio libero provident obcaecati a.",
                "date": "1539358547946",
                "username": "Richard Dawkins",
              },
              Object {
                "content": "Lorem ipsum dolor sit amet consectetur adipisicing elit. Architecto autem nam quae delectus maxime inventore odio libero provident obcaecati a.",
                "date": "1538288547946",
                "username": "Carl Sagan",
              },
              Object {
                "content": "Lorem ipsum dolor sit amet consectetur adipisicing elit. Architecto autem nam quae delectus maxime inventore odio libero provident obcaecati a.",
                "date": "1538998547946",
                "username": "Neil deGrasse Tyson",
              },
              Object {
                "content": "Lorem ipsum dolor sit amet consectetur adipisicing elit. Architecto autem nam quae delectus maxime inventore odio libero provident obcaecati a.",
                "date": "1539488547946",
                "username": "Christopher Hitchens",
              },
              Object {
                "content": "Lorem ipsum dolor sit amet consectetur adipisicing elit. Architecto autem nam quae delectus maxime inventore odio libero provident obcaecati a.",
                "date": "1538700547946",
                "username": "Richard Feynman",
              },
            ],
            "onInput": [Function],
            "onToggleComments": [Function],
            "showComments": true,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "onBtnClick": [Function],
            "showModal": false,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
